File: apps/web/src/lib/permissions.ts
import type { UserRole } from "@/types/auth";

export type WorkspaceRole = "owner" | "admin" | "member";

export type WorkspacePermission =
  | "members_manage"
  | "billing_manage"
  | "workspace_publish"
  | "workspace_edit"
  | "workspace_view_metrics"
  | "logs_view"
  | "workspace_admin"
  | "workspace_create"
  | "plan_view"
  | "plan_manage";

export type WorkspacePermissionMap = Record<WorkspacePermission, boolean>;

export const workspaceRolePermissions: Record<WorkspaceRole, WorkspacePermissionMap> = {
  owner: {
    members_manage: true,
    billing_manage: true,
    workspace_publish: true,
    workspace_edit: true,
    workspace_view_metrics: true,
    logs_view: true,
    workspace_admin: true,
    workspace_create: true,
    plan_view: true,
    plan_manage: true,
  },
  admin: {
    members_manage: true,
    billing_manage: false,
    workspace_publish: true,
    workspace_edit: true,
    workspace_view_metrics: true,
    logs_view: true,
    workspace_admin: false,
    workspace_create: true,
    plan_view: true,
    plan_manage: true,
  },
  member: {
    members_manage: false,
    billing_manage: false,
    workspace_publish: false,
    workspace_edit: true,
    workspace_view_metrics: true,
    logs_view: true,
    workspace_admin: false,
    workspace_create: true,
    plan_view: true,
    plan_manage: false,
  },
};

export function resolveWorkspaceRoleFromUser(userRole?: UserRole): WorkspaceRole {
  switch (userRole) {
    case "admin":
      return "owner";
    case "creator":
      return "admin";
    case "user":
    default:
      return "member";
  }
}

export function buildWorkspacePermissions(role: WorkspaceRole): WorkspacePermissionMap {
  return workspaceRolePermissions[role];
}

export function hasWorkspacePermission(
  permissions: Partial<WorkspacePermissionMap> | undefined,
  permission: WorkspacePermission
): boolean {
  return Boolean(permissions?.[permission]);
}

export function hasAnyWorkspacePermission(
  permissions: Partial<WorkspacePermissionMap> | undefined,
  ...required: WorkspacePermission[]
): boolean {
  return required.some((permission) => hasWorkspacePermission(permissions, permission));
}

File: apps/server/internal/service/permissions.go
package service

import (
	"strconv"
	"strings"

	"github.com/reverseai/server/internal/domain/entity"
)

const (
	PermissionMembersManage        = "members_manage"
	PermissionBillingManage        = "billing_manage"
	PermissionWorkspacePublish     = "workspace_publish"
	PermissionWorkspaceEdit        = "workspace_edit"
	PermissionWorkspaceViewMetrics = "workspace_view_metrics"
	PermissionLogsView             = "logs_view"
	PermissionWorkspaceAdmin       = "workspace_admin"
	PermissionWorkspaceCreate      = "workspace_create"
	PermissionWorkspaceDBAccess    = "workspace_db_access"
	PermissionPlanView             = "plan_view"
	PermissionPlanManage           = "plan_manage"
	PermissionConnectorsManage     = "connectors_manage"
)

var defaultWorkspaceRolePermissions = map[string]entity.JSON{
	"owner": {
		PermissionMembersManage:        true,
		PermissionBillingManage:        true,
		PermissionWorkspacePublish:     true,
		PermissionWorkspaceEdit:        true,
		PermissionWorkspaceViewMetrics: true,
		PermissionLogsView:             true,
		PermissionWorkspaceAdmin:       true,
		PermissionWorkspaceCreate:      true,
		PermissionWorkspaceDBAccess:    true,
		PermissionPlanView:             true,
		PermissionPlanManage:           true,
		PermissionConnectorsManage:     true,
	},
	"admin": {
		PermissionMembersManage:        true,
		PermissionWorkspacePublish:     true,
		PermissionWorkspaceEdit:        true,
		PermissionWorkspaceViewMetrics: true,
		PermissionLogsView:             true,
		PermissionWorkspaceCreate:      true,
		PermissionWorkspaceDBAccess:    true,
		PermissionPlanView:             true,
		PermissionPlanManage:           true,
		PermissionConnectorsManage:     true,
	},
	"member": {
		PermissionWorkspaceEdit:        true,
		PermissionWorkspaceViewMetrics: true,
		PermissionLogsView:             true,
		PermissionWorkspaceCreate:      true,
		PermissionWorkspaceDBAccess:    false,
		PermissionPlanView:             true,
		PermissionPlanManage:           false,
		PermissionConnectorsManage:     false,
	},
}

func hasPermission(permissions entity.JSON, key string) bool {
	if permissions == nil {
		return false
	}
	raw, ok := permissions[key]
	if !ok {
		return false
	}
	switch value := raw.(type) {
	case bool:
		return value
	case int:
		return value != 0
	case int64:
		return value != 0
	case float64:
		return value != 0
	case string:
		parsed, err := strconv.ParseBool(strings.TrimSpace(value))
		if err == nil {
			return parsed
		}
		return strings.TrimSpace(value) != ""
	default:
		return false
	}
}
